### ISSUE TITLE
twisted.conch.ssh.userauth.SSHUserAuthClient usage of preferredOrder unclear

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6850]
|<img alt="veloutin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| veloutin reported|
|-|-|
|Trac ID|trac#6850|
|Type|enhancement|
|Created|2013-11-22 15:43:09Z|
|Branch|https://github.com/twisted/twisted/tree/clarify-conch-preferredOrder-6850|

twisted.conch.ssh.userauth.SSHUserAuthClient documentation specifies the instance var _preferredOrder_ as "a list of authentication methods we support, in order of preference. The client will try authentication methods in this order, making callbacks for information when necessary. (type: list)".

In reality, the list of attempted authentications are the ones provided by the remote server in the _MSG_USERAUTH_FAILURE_ payload. Methods that are specified in _preferredOrder_ will be tried first, but others will still be attempted afterwards.

I see two options for this:
 * Change the docstring to reflect the behavior. (e.g "a list of preferred authentication methods. The client will try these authentication methods first if they are supported by the server, making callbacks for information when necessary. (type: list)".
 * Change ssh_USERAUTH_FAILURE to discard methods that are not in _preferredOrder_.

Attachments:

* [tw6850.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/89d/89d72f557a3a9a0cf2506fb98071d51a15977227/4f43d2d840082cdeb62f7c175938cd61c5d7a21a.diff) (1024 bytes) - added by veloutin on 2013-11-25 19:00:10Z - Docstring fix for twisted.conch.ssh.userauth.SSHUserAuthClient

<details><summary>Searchable metadata</summary>

```
trac-id__6850 6850
type__enhancement enhancement
reporter__veloutin veloutin
priority__normal normal
milestone__ 
branch__branches_clarify_conch_preferredOrder_6850 branches/clarify-conch-preferredOrder-6850
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1385134989000000 1385134989000000
changetime__1389529841000000 1389529841000000
version__None None
owner__ 
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="veloutin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|veloutin commented|
|-|-|

<exarkun> veloutin: Documentation should be corrected to describe the actual implementation except where there is a published standard that the implementation disagrees with.


The actual implementation is to try all authentication methods supported by the server, while giving priority to the ones in the preferredOrder list.

See ssh_USERAUTH_FAILURE:
```
    def ssh_USERAUTH_FAILURE(self, packet):
        """
        We received a MSG_USERAUTH_FAILURE.  Payload::
            string methods
            byte partial success

        If partial success is C{True}, then the previous method succeeded but is
        not sufficent for authentication. C{methods} is a comma-separated list
        of accepted authentication methods.

        We sort the list of methods by their position in C{self.preferredOrder},
        removing methods that have already succeeded. We then call
        C{self.tryAuth} with the most preferred method.

        @param packet: the L{MSG_USERAUTH_FAILURE} payload.
        @type packet: C{str}

        @return: a L{defer.Deferred} that will be callbacked with C{None} as
            soon as all authentication methods have been tried, or C{None} if no
            more authentication methods are available.
        @rtype: C{defer.Deferred} or C{None}
        """
        canContinue, partial = getNS(packet)
        partial = ord(partial)
        if partial:
            self.authenticatedWith.append(self.lastAuth)

        def orderByPreference(meth):
            """
            Invoked once per authentication method in order to extract a
            comparison key which is then used for sorting.

            @param meth: the authentication method.
            @type meth: C{str}

            @return: the comparison key for C{meth}.
            @rtype: C{int}
            """
            if meth in self.preferredOrder:
                return self.preferredOrder.index(meth)
            else:
                # put the element at the end of the list.
                return len(self.preferredOrder)

        canContinue = sorted([meth for meth in canContinue.split(',')
                              if meth not in self.authenticatedWith],
                             key=orderByPreference)

        log.msg('can continue with: %s' % canContinue)
        return self._cbUserauthFailure(None, iter(canContinue))
```

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_4"></a>|@hynek commented|
|-|-|

(In [41209]) Branching to clarify-conch-preferredOrder-6850.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_5"></a>|@hynek set status to `closed`|
|-|-|

(In [41213]) Merge clarify-conch-preferredOrder-6850

Author: veloutin
Reviewer: hynek
Fixes: [#6850](https://github.com/twisted/twisted/issues/6850)

Clarify the meaning of the preferredOrder ivar of
twisted.conch.ssh.userauth.SSHUserAuthClient.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_6"></a>|@hynek commented|
|-|-|

Thanks for this!

Next time please consider adding a topfile. :)
